### PR TITLE
Adaptive Telemetry Instrumentation

### LINKED ISSUE #1855: [PORT] Adaptive Telemetry Instrumentation
> Port this change from botbuilder-dotnet/master branch:
https://github.com/microsoft/botbuilder-dotnet/pull/3491

Addressed #2406 

Adds the P0 telemetry for adaptive - covering dialog start / cancel / complete, triggers, actions and adaptive recognizers.

# Changed projects
* AdaptiveExpressions
* Microsoft.Bot.Builder.Dialogs.Adaptive.Testing
* Microsoft.Bot.Builder.Dialogs.Adaptive
* Microsoft.Bot.Builder.Dialogs
* integration
* AdaptiveExpressions.Tests
* Microsoft.Bot.Builder.Dialogs.Adaptive.Tests

[adaptive,expression]

### LINKED ISSUE #2051: [PORT] Fixes for QnA / Recognizer Adaptive telemetry
> Port this change from botbuilder-dotnet/master branch:
https://github.com/microsoft/botbuilder-dotnet/pull/3715

 - Adds UseTelemetry extension method for DiaogManager which adds IBotTelemetryClient to turn state for use by adaptive recognizers.
 - Adaptive recognizers will now use the IBotTelemetryClient on the turn state unless the TelemetryClient property on the recognizer has been set.
 - Adds LogPersonalInformation property to QnAMaker dialogs and LUIS / QnA adaptive recognizers. Adaptive LUIS / QnA will look in settings.telemetry.logPersonalInformation by default.
 - Ensures QnAMaker client is always passed a TelemetryClient if set, along with the LogPersonalInformation flag (Fixes #3472)
 - LUIS Adaptive recognizer: Ensure telemetry properties passed in can override default properties

@cwhitten @vishwacsena Once these changes have been approved, I can raise an issue for composer to detail the changes needed in the runtime for telemetry.

# Changed projects
* Microsoft.Bot.Builder.AI.QnA
* Microsoft.Bot.Builder.Dialogs.Adaptive
* Microsoft.Bot.Builder.Dialogs

[adaptive]

### PR DESCRIPTION
**Pull request by songcs**
Fixes #1855 
Fixes #2051 

## Description
- Adds telemetry for adaptive, covering dialog start / cancel / complete, triggers, actions and adaptive recognizers.
- Adds useTelemetry extension method for DiaogManager which adds BotTelemetryClient to turn state for use by adaptive recognizers.
- Adaptive recognizers will now use the BotTelemetryClient on the turn state unless the telemetryClient property on the recognizer has been set.
- Adds logPersonalInformation property to QnAMaker dialogs and LUIS  adaptive recognizers. Adaptive LUIS will look in settings.telemetry.logPersonalInformation by default.

## Specific Changes

- botbuilder-ai
- botbuilder-dialogs
- botbuilder-dialogs-adaptive

### CODE REVIEW

**Comment by luhan2017** [comment]
@garypretty , could you help review this PR? we are not sure how to test it, could you provide some guidance?

File: libraries/botbuilder-dialogs-adaptive/src/luis/luisAdaptiveRecognizer.ts [hunk 1]
```diff
@@ -94,4 +91,59 @@ export class LuisAdaptiveRecognizer extends Configurable implements Recognizer {
...
+     * These properties are logged when the recognizer is called.
+     * @param recognizerResult Last activity sent from user.
+     * @param telemetryProperties Additional properties to be logged to telemetry with the LuisResult event.
+     * @param dialogContext Dialog context.
+     * @returns A dictionary that is sent as properties to BotTelemetryClient.trackEvent method for the LuisResult event.
+     */
+    protected  fillRecognizerResultTelemetryProperties(recognizerResult: RecognizerResult, telemetryProperties: { [key: string]: string }, dialogContext: DialogContext): { [key: string]: string } {
+        var logPersonalInfoVnE = this.logPersonalInformation.tryGetValue(dialogContext.state);
+        var logPersonalInfo = logPersonalInfoVnE.value;
+        var error = logPersonalInfoVnE.error;
```
Comment (chon219) [inline libraries/botbuilder-dialogs-adaptive/src/luis/luisAdaptiveRecognizer.ts:105]: Just ignore error values that has never been used.
const { logPersonalInfo } = this.logPersonalInformation.tryGetValue(dialogContext.state);

File: libraries/botbuilder-dialogs-adaptive/src/luis/luisAdaptiveRecognizer.ts [hunk 2]
```diff
@@ -94,4 +91,59 @@ export class LuisAdaptiveRecognizer extends Configurable implements Recognizer {
...
+
+        var applicationIdVnE = this.applicationId.tryGetValue(dialogContext.state);
+        var applicationId = applicationIdVnE.value;
+        var error2 = applicationIdVnE.error;
```
Comment (chon219) [inline libraries/botbuilder-dialogs-adaptive/src/luis/luisAdaptiveRecognizer.ts:109]: const { applicationId } = this.applicationId.tryGetValue(dialogContext.state);

File: libraries/botbuilder-dialogs-adaptive/src/luis/luisAdaptiveRecognizer.ts [hunk 3]
```diff
@@ -94,4 +91,59 @@ export class LuisAdaptiveRecognizer extends Configurable implements Recognizer {
...
+        
+        const topLuisIntent: string = LuisRecognizer.topIntent(recognizerResult);
+        const intentScore: number = (recognizerResult.intents[topLuisIntent] && 'score' in recognizerResult.intents[topLuisIntent]) ?
+            recognizerResult.intents[topLuisIntent].score : 0;
```
Comment (chon219) [inline libraries/botbuilder-dialogs-adaptive/src/luis/luisAdaptiveRecognizer.ts:113]: const intentScore: number = (recognizerResult.intents[topLuisIntent] && recognizerResult.intents[topLuisIntent].score) || 0;

File: libraries/botbuilder-dialogs-adaptive/src/luis/luisAdaptiveRecognizer.ts [hunk 4]
```diff
@@ -94,4 +91,59 @@ export class LuisAdaptiveRecognizer extends Configurable implements Recognizer {
...
+        properties[LuisTelemetryConstants.intentProperty] = topLuisIntent;
+        properties[LuisTelemetryConstants.intentScoreProperty] = intentScore.toString();
+        properties[LuisTelemetryConstants.fromIdProperty] = dialogContext.context.activity.from.id;
+
+        if (recognizerResult.sentiment) {
+            if (recognizerResult.sentiment.label) {
+                properties[LuisTelemetryConstants.sentimentLabelProperty] = recognizerResult.sentiment.label;
+            }
+
+            if (recognizerResult.sentiment.score) {
```
Comment (chon219) [inline libraries/botbuilder-dialogs-adaptive/src/luis/luisAdaptiveRecognizer.ts:127]: Can score be 0? which might be falsy.

File: libraries/botbuilder-dialogs-adaptive/src/recognizers/crossTrainedRecognizerSet.ts [hunk 5]
```diff
@@ -12,23 +12,28 @@ import { Recognizer } from './recognizer';
...
+            return recognizer.recognize(dialogContext, activity, telemetryProperties, telemetryMetrics);
         }));
+        
+        var result = this.processResults(results);
+        this.trackRecognizerResult(dialogContext, 'CrossTrainedRecognizerSetResult', this.fillRecognizerResultTelemetryProperties(result, telemetryProperties),telemetryMetrics);
+        return result;
+        
+    }
 
+    private processResults(results: RecognizerResult[]): RecognizerResult{
```
Comment (chon219) [inline libraries/botbuilder-dialogs-adaptive/src/recognizers/crossTrainedRecognizerSet.ts:36]: Add space before {

File: libraries/botbuilder-dialogs-adaptive/src/recognizers/crossTrainedRecognizerSet.ts [hunk 6]
```diff
@@ -12,23 +12,28 @@ import { Recognizer } from './recognizer';
...
                 throw new Error('This recognizer requires that each recognizer in the set have an id.');
             }
         }
 
         const results = await Promise.all(this.recognizers.map((recognizer: Recognizer): Promise<RecognizerResult> => {
-            return recognizer.recognize(dialogContext, activity);
+            return recognizer.recognize(dialogContext, activity, telemetryProperties, telemetryMetrics);
         }));
+        
+        var result = this.processResults(results);
```
Comment (chon219) [inline libraries/botbuilder-dialogs-adaptive/src/recognizers/crossTrainedRecognizerSet.ts:30]: const

File: libraries/botbuilder-dialogs-adaptive/src/recognizers/multiLanguageRecognizer.ts [hunk 7]
```diff
@@ -35,7 +33,10 @@ export class MultiLanguageRecognizer implements Recognizer {
             const option = policy[i];
             if (this.recognizers.hasOwnProperty(option)) {
                 const recognizer = this.recognizers[option];
-                return await recognizer.recognize(dialogContext, activity);
+                var result = await recognizer.recognize(dialogContext, activity, telemetryProperties, telemetryMetrics);
```
Comment (chon219) [inline libraries/botbuilder-dialogs-adaptive/src/recognizers/multiLanguageRecognizer.ts:36]: const

File: libraries/botbuilder-dialogs-adaptive/src/recognizers/recognizer.ts [hunk 8]
```diff
@@ -5,12 +5,91 @@
...
+
+    /**
+     * To recognize intents and entities in a users utterance.
+     *
+     * @param dialogContext Dialog Context.
+     * @param activity Activity.
+     * @param telemetryProperties Additional properties to be logged to telemetry with event.
+     * @param telemetryMetrics Additional metrics to be logged to telemetry with event.
+     */
+    public recognize(dialogContext: DialogContext, activity: Partial<Activity>, telemetryProperties?: { [key: string]: string }, telemetryMetrics?: { [key: string]: number }): Promise<RecognizerResult>{
```
Comment (chon219) [inline libraries/botbuilder-dialogs-adaptive/src/recognizers/recognizer.ts:48]: space before {

File: libraries/botbuilder-dialogs-adaptive/src/recognizers/recognizer.ts [hunk 9]
```diff
@@ -5,12 +5,91 @@
...
+    }
+
+    /**
+     * Uses the RecognizerResult to create a list of propeties to be included when tracking the result in telemetry.
+     * @param recognizerResult Recognizer Result.
+     * @param telemetryProperties A list of properties to append or override the properties created using the RecognizerResult.
+     * @param dialogContext Dialog Context.
+     * @returns A dictionary that can be included when calling the TrackEvent method on the TelemetryClient.
+     */
+    protected fillRecognizerResultTelemetryProperties(recognizerResult: RecognizerResult, telemetryProperties: { [key: string]: string }, dialogContext?: DialogContext): { [key: string]: string }{
```
Comment (chon219) [inline libraries/botbuilder-dialogs-adaptive/src/recognizers/recognizer.ts:59]: space before {

File: libraries/botbuilder-dialogs-adaptive/src/recognizers/recognizer.ts [hunk 10]
```diff
@@ -5,12 +5,91 @@
...
+
+        var properties: { [key: string]: string } = {
+            'Text': recognizerResult.text ,
+            'AlteredText': recognizerResult.alteredText ,
+            'TopIntent': Object.entries(recognizerResult.intents).length > 0 ? intent : undefined ,
+            'TopIntentScore': Object.entries(recognizerResult.intents).length > 0 ? score.toString() : undefined ,
+            'Intents': Object.entries(recognizerResult.intents).length > 0 ? JSON.stringify(recognizerResult.intents) : undefined , 
+            'Entities': recognizerResult.entities ? recognizerResult.entities.toString() : undefined ,
+            'AdditionalProperties': recognizerResult.properties ? (Object.entries(recognizerResult.properties).length > 0 ? JSON.stringify(recognizerResult.properties) : undefined ) : undefined
+        };
```
Comment (chon219) [inline libraries/botbuilder-dialogs-adaptive/src/recognizers/recognizer.ts:71]: const

File: libraries/botbuilder-dialogs-adaptive/src/recognizers/recognizer.ts [hunk 11]
```diff
@@ -5,12 +5,91 @@
...
+        this._telemetryClient = client ? client : new NullTelemetryClient();
+    }
+
+    /**
+     * To recognize intents and entities in a users utterance.
+     *
+     * @param dialogContext Dialog Context.
+     * @param activity Activity.
+     * @param telemetryProperties Additional properties to be logged to telemetry with event.
+     * @param telemetryMetrics Additional metrics to be logged to telemetry with event.
```
Comment (chon219) [inline libraries/botbuilder-dialogs-adaptive/src/recognizers/recognizer.ts:46]: Initial uppercase

File: libraries/botbuilder-dialogs-adaptive/src/recognizers/recognizer.ts [hunk 12]
```diff
@@ -5,12 +5,91 @@
...
+     */
+    public recognize(dialogContext: DialogContext, activity: Partial<Activity>, telemetryProperties?: { [key: string]: string }, telemetryMetrics?: { [key: string]: number }): Promise<RecognizerResult>{
+        throw new Error('Please implement recognize function.');
+    }
+
+    /**
+     * Uses the RecognizerResult to create a list of propeties to be included when tracking the result in telemetry.
+     * @param recognizerResult Recognizer Result.
+     * @param telemetryProperties A list of properties to append or override the properties created using the RecognizerResult.
+     * @param dialogContext Dialog Context.
```
Comment (chon219) [inline libraries/botbuilder-dialogs-adaptive/src/recognizers/recognizer.ts:56]: Initial upper case.

File: libraries/botbuilder-dialogs-adaptive/src/recognizers/recognizer.ts [hunk 13]
```diff
@@ -5,12 +5,91 @@
...
+     * @returns A dictionary that can be included when calling the TrackEvent method on the TelemetryClient.
+     */
+    protected fillRecognizerResultTelemetryProperties(recognizerResult: RecognizerResult, telemetryProperties: { [key: string]: string }, dialogContext?: DialogContext): { [key: string]: string }{
+        
+        const {intent, score} = getTopScoringIntent(recognizerResult);
```
Comment (chon219) [inline libraries/botbuilder-dialogs-adaptive/src/recognizers/recognizer.ts:61]: spaces around variables in {}

File: libraries/botbuilder-dialogs-adaptive/src/recognizers/recognizer.ts [hunk 14]
```diff
@@ -5,12 +5,91 @@
...
+
+        var properties: { [key: string]: string } = {
+            'Text': recognizerResult.text ,
+            'AlteredText': recognizerResult.alteredText ,
+            'TopIntent': Object.entries(recognizerResult.intents).length > 0 ? intent : undefined ,
+            'TopIntentScore': Object.entries(recognizerResult.intents).length > 0 ? score.toString() : undefined ,
+            'Intents': Object.entries(recognizerResult.intents).length > 0 ? JSON.stringify(recognizerResult.intents) : undefined , 
+            'Entities': recognizerResult.entities ? recognizerResult.entities.toString() : undefined ,
+            'AdditionalProperties': recognizerResult.properties ? (Object.entries(recognizerResult.properties).length > 0 ? JSON.stringify(recognizerResult.properties) : undefined ) : undefined
```
Comment (chon219) [inline libraries/botbuilder-dialogs-adaptive/src/recognizers/recognizer.ts:70]: We don't have a property named "properties" in recognizer result.

File: libraries/botbuilder-dialogs-adaptive/src/recognizers/recognizer.ts [hunk 15]
```diff
@@ -5,12 +5,91 @@
...
+        // Additional Properties can override "stock" properties.
+        if (telemetryProperties){
+            return Object.assign({}, properties, telemetryProperties);
+        }
+        return properties;
+    }
+
+    protected trackRecognizerResult(dialogContext: DialogContext, eventName: string, telemetryProperties?: { [key: string]: string }, telemetryMetrics?: { [key: string]: number }){
+        if (this.telemetryClient instanceof NullTelemetryClient) {
+            var turnStateTelemetryClient = dialogContext.context.turnState.get(telemetryClientKey);
```
Comment (chon219) [inline libraries/botbuilder-dialogs-adaptive/src/recognizers/recognizer.ts:83]: const

File: libraries/botbuilder-dialogs-adaptive/src/recognizers/recognizer.ts [hunk 16]
```diff
@@ -5,12 +5,91 @@
...
+
+
+        // Additional Properties can override "stock" properties.
+        if (telemetryProperties){
+            return Object.assign({}, properties, telemetryProperties);
+        }
+        return properties;
+    }
+
+    protected trackRecognizerResult(dialogContext: DialogContext, eventName: string, telemetryProperties?: { [key: string]: string }, telemetryMetrics?: { [key: string]: number }){
```
Comment (chon219) [inline libraries/botbuilder-dialogs-adaptive/src/recognizers/recognizer.ts:81]: space before {

File: libraries/botbuilder-dialogs-adaptive/src/recognizers/recognizer.ts [hunk 17]
```diff
@@ -5,12 +5,91 @@
...
+        if (this.telemetryClient instanceof NullTelemetryClient) {
+            var turnStateTelemetryClient = dialogContext.context.turnState.get(telemetryClientKey); 
+            this.telemetryClient = turnStateTelemetryClient ? turnStateTelemetryClient: this.telemetryClient;
```
Comment (chon219) [inline libraries/botbuilder-dialogs-adaptive/src/recognizers/recognizer.ts:84]: this.telemetryClient = turnStateTelemetryClient || this.telemetryClient;

File: libraries/botbuilder-dialogs-adaptive/src/recognizers/recognizer.ts [hunk 18]
```diff
@@ -5,12 +5,91 @@
...
+        }
+        this.telemetryClient.trackEvent(
+            {
```
Comment (chon219) [inline libraries/botbuilder-dialogs-adaptive/src/recognizers/recognizer.ts:87]: brackets in same line.

File: libraries/botbuilder-dialogs-adaptive/src/recognizers/recognizer.ts [hunk 19]
```diff
@@ -5,12 +5,91 @@
...
+            'TopIntent': Object.entries(recognizerResult.intents).length > 0 ? intent : undefined ,
+            'TopIntentScore': Object.entries(recognizerResult.intents).length > 0 ? score.toString() : undefined ,
+            'Intents': Object.entries(recognizerResult.intents).length > 0 ? JSON.stringify(recognizerResult.intents) : undefined , 
+            'Entities': recognizerResult.entities ? recognizerResult.entities.toString() : undefined ,
+            'AdditionalProperties': recognizerResult.properties ? (Object.entries(recognizerResult.properties).length > 0 ? JSON.stringify(recognizerResult.properties) : undefined ) : undefined
+        };
+
+
+        // Additional Properties can override "stock" properties.
+        if (telemetryProperties){
```
Comment (chon219) [inline libraries/botbuilder-dialogs-adaptive/src/recognizers/recognizer.ts:75]: space

File: libraries/botbuilder-dialogs-adaptive/src/telemetryExtensions.ts [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { DialogManager } from 'botbuilder-dialogs';
+import { BotTelemetryClient } from 'botbuilder-core';
+
+
+export const telemetryClientKey = Symbol('telemetryClient');
```
Comment (chon219) [inline libraries/botbuilder-dialogs-adaptive/src/telemetryExtensions.ts:13]: documentation

File: libraries/botbuilder-dialogs-adaptive/src/telemetryExtensions.ts [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+
+export class TelemetryExtensions {
+    /**
+     * Configures the telemetry client to use.
+     * @param dialogManager DialogManager to configure.
+     * @param telemetryClient BotTelemetryClient instance to use.
```
Comment (chon219) [inline libraries/botbuilder-dialogs-adaptive/src/telemetryExtensions.ts:19]: Initial upper case.

**Comment by chon219** [comment]
BTW, add documentation for all public methods touched.

File: libraries/botbuilder-dialogs-adaptive/src/actions/actionScope.ts [hunk 22]
```diff
@@ -143,9 +143,17 @@ export class ActionScope<O extends object = {}> extends Dialog<O> implements Dia
             return await dc.endDialog();
         }
 
-        const actionId = this.actions[offset].id;
+        var action = this.actions[offset];
+        var actionName = action.constructor.name;
```
Comment (chon219) [inline libraries/botbuilder-dialogs-adaptive/src/actions/actionScope.ts:147]: const

File: libraries/botbuilder-dialogs-adaptive/src/luis/luisAdaptiveRecognizer.ts [hunk 23]
```diff
@@ -73,11 +67,14 @@ export class LuisAdaptiveRecognizer extends Configurable implements Recognizer {
...
             endpoint: this.endpoint.getValue(dcState),
             endpointKey: this.endpointKey.getValue(dcState)
-        }
+        };
 
         // Create and call wrapper
         const wrapper = new LuisRecognizer(application, this.recognizerOptions(dialogContext));
-        return await wrapper.recognize(context);
+
+        var result = await wrapper.recognize(context);
```
Comment (chon219) [inline libraries/botbuilder-dialogs-adaptive/src/luis/luisAdaptiveRecognizer.ts:75]: const

File: libraries/botbuilder-ai/src/index.ts [hunk 24]
```diff
@@ -11,7 +11,8 @@ export * from './geographyV2';
 export * from './instanceData';
 export * from './intentData';
 export * from './luisRecognizer';
+export * from './luisTelemetryConstants';
```
Comment (stevengum) [inline libraries/botbuilder-ai/src/index.ts:14]: This is parity with .NET: https://github.com/microsoft/botbuilder-dotnet/blob/20b1539f4eea4dea0bf184ca6e75b6ae6f033748/libraries/Microsoft.Bot.Builder.AI.LUIS/LuisTelemetryConstants.cs#L9

File: libraries/botbuilder-dialogs/src/dialogManager.ts [hunk 25]
```diff
@@ -57,6 +57,7 @@ export class DialogManager extends Configurable {
     private readonly _dialogStateProperty: string;
     private readonly _initialTurnState: TurnContextStateCollection = new TurnContextStateCollection();
 
+
```
Comment (stevengum) [inline libraries/botbuilder-dialogs/src/dialogManager.ts:60]: Undo this change?

File: libraries/botbuilder-dialogs-adaptive/src/telemetryExtensions.ts [hunk 26]
```diff
@@ -0,0 +1,32 @@
...
+
+/**
+ * The key to get or set telemetry client from turn state.
+ */
+export const telemetryClientKey = Symbol('telemetryClient');
+
+/**
+ * Extension methods for telemetry.
+ */
+export class TelemetryExtensions {
```
Comment (stevengum) [inline libraries/botbuilder-dialogs-adaptive/src/telemetryExtensions.ts:20]: The current naming protocol is to name the class with the Class/Type it is extending, in this case it would be "DialogManagerExtensions"

File: libraries/botbuilder-dialogs-adaptive/src/index.ts [hunk 27]
```diff
@@ -29,6 +29,7 @@ export * from './resourceExtensions';
 export * from './schemaHelper';
 export * from './selectors';
 export * from './skillExtensions';
+export * from './telemetryExtensions';
```
Comment (stevengum) [inline libraries/botbuilder-dialogs-adaptive/src/index.ts:32]: Please see previous comment about the class name and update the filename accordingly.

File: libraries/botbuilder-dialogs/src/dialogManager.ts (hunk 25 above)
Comment (songcs) [reply libraries/botbuilder-dialogs/src/dialogManager.ts]: Modification has been undone.

File: libraries/botbuilder-dialogs-adaptive/src/telemetryExtensions.ts (hunk 26 above)
Comment (chon219) [reply libraries/botbuilder-dialogs-adaptive/src/telemetryExtensions.ts:20]: @stevengum We have several helper classes (e.g. SkillExtensions, LanguageGenerationExtensions) to extend DialogManager, is it required to rename all of them to DialogManagerExtensions?

Comment (stevengum) [reply libraries/botbuilder-dialogs-adaptive/src/telemetryExtensions.ts:20]: In .NET, the extension methods and IntelliSense make it easy for developers to use these static helpers, so the class names haven't been as important. That said, @gabog and @johnataylor have indicated that the correct naming pattern is to use what is being extended as the class name.

From a porting perspective, using a singular `DialogManagerExtensions` helps in TypeScript, Python and Java as the developer only has to import `DialogManagerExtensions` and then use the appropriate static helper, versus having several different imports to access 1 or 2 helpers per import.

We may not fix this for the existing extensions in .NET (I'm not sure if this breaks binary or semver compat, though I don't believe it does), but going forward the consensus is to use `<ExtendedClass>Extensions` AFAIK.

Comment (chon219) [reply libraries/botbuilder-dialogs-adaptive/src/telemetryExtensions.ts:20]: We can combine these helper classes into one in _botbuilder-dialogs-adaptive_ named `DialogManagerExtensions`. However, if we want to extend `DialogManager` from different packages, we may need to create multiple `DialogManagerExtensions` for each package, would that possibly an issue?

Comment (stevengum) [reply libraries/botbuilder-dialogs-adaptive/src/telemetryExtensions.ts:20]: It wouldn't because the ultimate reference name is defined by the consumer of the import:

```ts
import { DialogManagerExtensions as AdaptiveDMExtensions } from 'botbuilder-dialogs-adaptive';
import { DialogManagerExtensions } from 'botbuilder-dialogs-declarative';
import { DialogManagerExtensions as LUISDMExtensions }from 'botbuilder-ai';

```

Comment (chon219) [reply libraries/botbuilder-dialogs-adaptive/src/telemetryExtensions.ts:20]: So is it okay to merge `LanguageGeneratorExtensions`, `ResourceExtensions`, `SkillExtensions` and `TelemetryExtensions` into one `DialogManagerExtensions` ? As they are all helper classes extending DialogManager.

Comment (chon219) [reply libraries/botbuilder-dialogs-adaptive/src/telemetryExtensions.ts:20]: @Stevenic As we discussed before, we need to keep the name `TelemetryExtensions` as the c# sdk has been shipped already, right?

Comment (stevengum) [reply libraries/botbuilder-dialogs-adaptive/src/telemetryExtensions.ts:20]: FYI @steveinc @johnataylor @tomlm @gabog

I don't think the value gained here by having the same name for extension methods is comparable to having the same names for actual classes, constants and interfaces across languages.

In C#, a developer generally doesn't have to worry about importing the correct `static class` for extension methods because the methods are just "provided" via namespace and intellisense. The syntactic sugar makes it easy to consume extension methods (aka static helper methods) in C#.

If I'm `using Microsoft.Bot.Builder.Dialogs.Adaptive;`, then on my DialogManager instance I just call `.UseTelemetry()`.

___

In TS/JS I have to import the specific class (demonstrated in an [earlier comment](https://github.com/microsoft/botbuilder-js/pull/2513/files#r462874559)) and then reference the method I want:

```ts
const dm = new DialogManager(/* ... */);

import { LanguageGeneratorExtensions , SkillExtensions, TelemetryExtensions } from 'botbuilder-dialogs-adaptive';

TelemetryExtensions.useTelemetry(dm, telemetryClient);
SkillExtensions.useSkillClient(dm, skillClient);
LanguageGeneratorExtensions.useLanguageGeneration(dm, 'myMain.lg');
```

If we decide we want to have separate files, then the extension methods should just be turned into standalone static functions without a class.

```ts
const dm = new DialogManager(/* ... */);

import { useTelemetry , useSkillClient, useLanguageGeneration } from 'botbuilder-dialogs-adaptive';

useTelemetry(dm, telemetryClient);
useSkillClient(dm, skillClient);
useLanguageGeneration(dm, 'myMain.lg');
```

___

If the consensus is to keep stay with `TelemetryExtensions` (aka different classes, different files) instead of using `DialogManagerExtensions` (aka single class, single file), then I'm okay with doing it; but from a design POV it's just added superfluous work on our developers.

Comment (chon219) [reply libraries/botbuilder-dialogs-adaptive/src/telemetryExtensions.ts:20]: I think it might be better to export standalone static functions instead of extension classes.

Comment (chon219) [reply libraries/botbuilder-dialogs-adaptive/src/telemetryExtensions.ts:20]: I've asked @songcs to export `useTelemetry()` as a standalone static function and keep the filename unchanged.
If it is okay to implement extension methods in this way, I would send out another PR to rename the other extension methods as we don't need to wrap these methods in a class which is useless in javascript.

Comment (stevengum) [reply libraries/botbuilder-dialogs-adaptive/src/telemetryExtensions.ts]: @chon219 that sounds good to me, please send out another PR to rename/refactor the other extension methods.
